docs/sbft-formal-model/replica.i.dfy

```diff
@@ -35,35 +38,45 @@ module Replica {
       && forall x | x in cs :: && cs[x].payload.Commit?
                                && c.clusterConfig.IsReplica(cs[x].sender)
   }
-
-  type PrePreparesRcvd = imap<SequenceID, Option<Network.Message<Message>>>
-  predicate PrePreparesRcvdWF(prePreparesRcvd:PrePreparesRcvd) {
-    && FullImap(prePreparesRcvd)
-    && (forall x | x in prePreparesRcvd && prePreparesRcvd[x].Some? :: prePreparesRcvd[x].value.payload.PrePrepare?)
+  function EmptyCommitProofSet() : CommitProofSet {
+    map[]
   }
 
   // The Working Window data structure. Here Replicas keep the PrePrepare from the Primary
   // and the votes from all peers. Once a Client Operation is committed by a given Replica
   // to a specific Sequence ID (the Replica has collected the necessary quorum of votes from
   // peers) the Client Operation is inserted in the committedClientOperations as appropriate.
   datatype WorkingWindow = WorkingWindow(
-    committedClientOperations:imap<SequenceID, Option<OperationWrapper>>,
-    prePreparesRcvd:PrePreparesRcvd,
-    preparesRcvd:imap<SequenceID, PrepareProofSet>,
-    commitsRcvd:imap<SequenceID, CommitProofSet>
+    committedClientOperations:map<SequenceID, Option<OperationWrapper>>,
+    prePreparesRcvd:map<SequenceID, Option<Network.Message<Message>>>,
+    preparesRcvd:map<SequenceID, PrepareProofSet>,
+    commitsRcvd:map<SequenceID, CommitProofSet>,
+    lastStableCheckpoint:SequenceID
   ) {
+    function getActiveSequenceIDs(c:Constants) : set<SequenceID> 
+      requires c.WF()
+    {
+      set seqID | lastStableCheckpoint <= seqID < lastStableCheckpoint + c.clusterConfig.workingWindowSize //TODO: refactor to finite sets/maps
+    }
     predicate WF(c:Constants)
       requires c.WF()
     {
-      && FullImap(committedClientOperations)
-      && FullImap(preparesRcvd)
-      && FullImap(commitsRcvd)
-      && PrePreparesRcvdWF(prePreparesRcvd)
+      && committedClientOperations.Keys == getActiveSequenceIDs(c)
+      && preparesRcvd.Keys == getActiveSequenceIDs(c)
+      && commitsRcvd.Keys == getActiveSequenceIDs(c)
+      && prePreparesRcvd.Keys == getActiveSequenceIDs(c)
+      && (forall x | x in prePreparesRcvd && prePreparesRcvd[x].Some? :: prePreparesRcvd[x].value.payload.PrePrepare?)
       && (forall seqID | seqID in preparesRcvd :: PrepareProofSetWF(c, preparesRcvd[seqID]))
       && (forall seqID | seqID in commitsRcvd :: CommitProofSetWF(c, commitsRcvd[seqID]))
     }
+    function ShiftWorkingWindow<T>(c:Constants, m:map<SequenceID,T>, empty:T) : map<SequenceID,T> 
```

Comment:
or simply, `Shift()`, as this function operates on the `workingWindow` structure?

Reply:
Yes, I will change it, now when I look at it, it is a bit too wordy in the usage :)